cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -241,10 +263,123 @@ private static ReportGenerationRequest getReportRequest(Location reportIdDir) th
+        return;
+      }
+      ReportStatus status;
+      try {
+        status = getReportStatus(reportIdDir);
+      } catch (Exception e) {
+        LOG.error("Failed to get the status of the report {}", reportId, e);
+        responder.sendError(500, String.format("Failed to get the status of the report %s becasue of error: %s",
+                                               reportId, e.getMessage()));
+        return;
+      }
+      // if the report is already deleted, deleting the report again is not allowed
+      if (ReportStatus.DELETED.equals(status)) {
+        responder.sendError(404, String.format("Report with id %s has already been deleted", reportId));
+        return;
+      }
+      Location deletedFile = reportIdDir.append(TO_BE_DELETED_FILE);
+      if (!deletedFile.createNew()) {
+        reportIdDir.delete();
+        responder.sendError(500, "Failed to delete " + reportId);
+        return;
+      }
+      // Save the time of when this report is marked as to-be-deleted in the _TO_BE_DELETED_FILE file
+      try (PrintWriter writer = new PrintWriter(new OutputStreamWriter(deletedFile.getOutputStream(),
+                                                                       StandardCharsets.UTF_8), true)) {
+        writer.write(Long.toString(System.currentTimeMillis()));
+      }
+      responder.sendStatus(200);
+    }
+
+    @POST
+    @Path("/reports/{report-id}/save")
+    public void saveReport(HttpServiceRequest request, HttpServiceResponder responder,
+                                @PathParam("report-id") String reportId) throws IOException {
+      Location reportIdDir = getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(reportId);
+      if (!reportIdDir.exists()) {
+        responder.sendError(404, String.format("Report with id %s does not exist.", reportId));
+        return;
+      }
+      ReportStatus status;
+      try {
+        status = getReportStatus(reportIdDir);
+      } catch (Exception e) {
+        LOG.error("Failed to get the status of the report {}", reportId, e);
+        responder.sendError(500, String.format("Failed to get the status of the report %s becasue of error: %s",
+                                               reportId, e.getMessage()));
+        return;
+      }
+      // only allow saving the report when its status is COMPLETED
+      switch (status) {
+        case COMPLETED:
+          break;
+        case DELETED:
+          responder.sendError(404, String.format("Report with id %s does not exist.", reportId));
+          return;
+        default:
+          responder.sendError(403, "Cannot save the report with status " + status);
+          return;
+      }
+      // if the report is already saved, saving the report again is not allowed
+      Location savedFile = reportIdDir.append(SAVED_FILE);
+      if (savedFile.exists()) {
+        responder.sendError(403, String.format("Report with id %s is already saved and cannot be saved again.",
```

Comment:
so we don't support updating description is that correct?

Reply:
i think we should support overwrite, there could be case where file is created and during write it fails and user wants to retry.

Reply:
the UI doesn't allow users to change the name and description once they save the report. For the case where the file is created but write operation failed, should we only allow updating the file if the content in the file is not a valid save request?

Reply:
implementation and what UI does are two separate things. i think we should relax this check.

Reply:
if the report is marked as DELETED, users won't see them in the UI, unless they keep a note of the report ID and call the REST API directly. It provides a grace period if the user is already reading the report while the report is being deleted, especially when we support sharing reports among users.